src/core/mac/mac.cpp

```diff
@@ -1571,8 +1571,53 @@ void Mac::ReceiveDoneTask(Frame *aFrame, otError aError)
     SuccessOrExit(error = aFrame->ValidatePsdu());
 
     aFrame->GetSrcAddr(srcaddr);
+    aFrame->GetDstAddr(dstaddr);
     neighbor = GetNetif().GetMle().GetNeighbor(srcaddr);
 
+    // Destination Address Filtering
+    switch (dstaddr.mLength)
+    {
+    case 0:
+        break;
+
+    case sizeof(ShortAddress):
+        aFrame->GetDstPanId(panid);
+        VerifyOrExit((panid == kShortAddrBroadcast || panid == mPanId) &&
+                     ((mRxOnWhenIdle && dstaddr.mShortAddress == kShortAddrBroadcast) ||
+                      dstaddr.mShortAddress == mShortAddress), error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
+
+#if OPENTHREAD_FTD
+
+        // Allow non link-local multicasts from neighbor routers if FFD
+        if (neighbor == NULL && dstaddr.mShortAddress == kShortAddrBroadcast &&
+            srcaddr.mLength == sizeof(ShortAddress) &&
+            GetNetif().GetMle().IsActiveRouter(srcaddr.mShortAddress) &&
```

Comment:
Is this check necessary?  I believe all neighbors in the valid state should be active routers.

Reply:
The IsActiveRouter function just checks the RLOC16 has a Child ID of 0, and is related to how I obtain the router object below (Otherwise a child RLOC would be accepted but the ```neighbor``` pointer would refer to the parent). I think I'm going to abstract this away, because it did make for an ugly bloated if condition, and I can also remove the preprocessor conditionals.